File: Binary-search/testingbinarysearch.c
// binary search algorithm
// 1. sort the array
// 2. find the middle element
// 3. if the middle element is the target, return the index
// 4. if the middle element is greater than the target, search the left half
// 5. if the middle element is less than the target, search the right half
// 6. repeat steps 2-5 until the target is found or the array is empty
#include <stdio.h>
#include <stdlib.h>
int binarySearch(int arr[], int target, int size);
int main()
{
    int arr[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    int target = 5;
    int size = sizeof(arr) / sizeof(arr[0]);
    int index = binarySearch(arr, target, size);
    if (index == -1)
    {
        printf("Target not found");
    }
    else
    {
        printf("Target found at index %d", index);
    }
    return 0;
}
int binarySearch(int arr[], int target, int size)
{
    int left = 0;
    int right = size - 1;
    while (left <= right)
    {
        int middle = (left + right) / 2;
        if (arr[middle] == target)
        {
            return middle;
        }
        else if (arr[middle] > target)
        {
            right = middle - 1;
        }
        else
        {
            left = middle + 1;
        }
    }
    return -1;
}

File: Binary-search/testingbinarysearch.js
//binary search algorithm
//binary search is a search algorithm that finds the position of a target value within a sorted array.
//binary search compares the target value to the middle element of the array.
//if they are not equal, the half in which the target cannot lie is eliminated and the search continues on the remaining half, again taking the middle element to compare to the target value, and repeating this until the target value is found.
//if the search ends with the remaining half being empty, the target is not in the array.
//binary search runs in logarithmic time in the worst case, making O(log n) comparisons, where n is the number of elements in the array, and therefore is much faster than linear search O(n) which makes at most O(n) comparisons, where n is the number of elements in the array.
//if the search space is reduced by half at each comparison, then the total number of comparisons will be the number of times one can halve n before getting a value that is less than or equal to one.
//that is if n is 16, then the maximum number of times one can halve 16 to get a value less than or equal to one is four.
//if n is 32, then the maximum number of times one can halve 32 to get a value less than or equal to one is five.
//this can be computed using the following formula:
//log base 2 of n equals the number of times one can halve n to get a value less than or equal to one.
const binarySearch = (arr, target) => {
  let start = 0;
  let end = arr.length - 1;
  let middle = Math.floor((start + end) / 2);
  while (arr[middle] !== target && start <= end) {
    if (target < arr[middle]) end = middle - 1;
    else start = middle + 1;
    middle = Math.floor((start + end) / 2);
  }
  return arr[middle] === target ? middle : -1;
};
console.log(binarySearch([2, 5, 6, 9, 13, 15, 28, 30], 103));
//output: -1
//Time complexity: O(log n)
//Space complexity: O(1)
